### PR TITLE
fix: fix incorrect parent for type reference of a CtParameter

### LINKED ISSUE #3671: parameter.getReference() would set parent of parameter type, and Spoon cannot get declaration of generic method by reference.
I'm working with Spoon 8.3.0, and I want to analyze some Java8 code.

Here is a snippet to reproduce the issue

```
        SpoonAPI spoon = new Launcher();
        spoon.addInputResource("/home/usr/TestGeneric.java");
        spoon.buildModel();

        Map<CtExecutableReference<?>, List<CtParameterReference<?>>> exe2Parameter =
                new HashMap<>();
        for (CtExecutable<?> cm : spoon.getModel().getRootPackage()
               .getElements(new TypeFilter<>(CtExecutable.class))) {

            List<CtParameterReference<?>> pList = new ArrayList<>();
            for(CtParameter<?> parameter : cm.getParameters()) {
                //getReference() change the parameter.getType().getParent()
                pList.add(parameter.getReference());
            }
            exe2Parameter.put(cm.getReference(), pList);
        }
        for (CtExecutableReference<?> exeRef : exe2Parameter.keySet()) {
            //Getting declaration of generic method would throw exception.
            System.out.println(exeRef.getExecutableDeclaration());
        }
```
TestGeneric.java
```
public class TestGeneric {
    public static class Cell<T> {
        public T val;
        public T getVal() { return val; }
        public void setVal(T val) { this.val = val;}
    }

    public static class Day {
        public int day;
        public Day() { }
        public Day(int val) {this.day = val;}
    }

    public void bar() {
        Cell<Day> c = new Cell<>();
        Day e = new Day();
        c.setVal(e);
    }
}
```

```
Exception in thread "main" spoon.SpoonException: Cannot resolve type erasure of the type parameter reference, which is not able to found it's declaration.
	at spoon.support.reflect.reference.CtTypeParameterReferenceImpl.getTypeErasure(CtTypeParameterReferenceImpl.java:180)
	at spoon.support.reflect.declaration.CtTypeImpl.isSameParameter(CtTypeImpl.java:752)
	at spoon.support.reflect.declaration.CtTypeImpl.hasSameParameters(CtTypeImpl.java:744)
	at spoon.support.reflect.declaration.CtTypeImpl.getMethod(CtTypeImpl.java:718)
	at spoon.support.reflect.reference.CtExecutableReferenceImpl.getCtExecutable(CtExecutableReferenceImpl.java:121)
	at spoon.support.reflect.reference.CtExecutableReferenceImpl.getExecutableDeclaration(CtExecutableReferenceImpl.java:113)
	at GetInvocation2.main(GetInvocation2.java:30)

```

Is this a bug, or I just need to avoid getting reference of method parameter before getting method declaration?

### PR DESCRIPTION
**Pull request by andrewbwogi**
close #3671

### CODE REVIEW

**Comment by monperrus** [comment]
Thanks @andrewbwogi for the work on this fix. Could you have a look at the CI failures?

**Comment by andrewbwogi** [comment]
This solution works but it hides an implicit difference between ordinary parameters and parameters in lambda expressions. Is it better to create a CtLambdaParameter class?

File: src/main/java/spoon/reflect/factory/ExecutableFactory.java [hunk 1]
```diff
@@ -79,7 +79,11 @@ public CtAnonymousExecutable createAnonymous(CtClass<?> target, CtBlock<Void> bo
 	public <T> CtParameterReference<T> createParameterReference(CtParameter<T> parameter) {
 		CtParameterReference<T> ref = factory.Core().createParameterReference();
 		ref.setSimpleName(parameter.getSimpleName());
-		ref.setType(parameter.getType());
+		if (parameter.getType() == null) {
+			ref.setType(parameter.getType());
+		} else {
+			ref.setType(parameter.getType().clone());
+		}
```
Comment (andrewbwogi) [inline src/main/java/spoon/reflect/factory/ExecutableFactory.java:86]: This solution works but it hides an implicit difference between ordinary parameters and parameters in lambda expressions. Is it better to create a CtLambdaParameter class?

**Comment by monperrus** [comment]
>  Is it better to create a CtLambdaParameter class?

It would probably be overkill.

Do you mean that we don't have a fully backward compatible fix for now?

**Comment by andrewbwogi** [comment]
The fix only expands information in the AST so it should be backwards compatible.

**Comment by monperrus** [comment]
Great, thanks @andrewbwogi !

In Spoon, we always have a one line contract in natural language for each test:

    // contract: after adding an element in a list, the size is incremented

Could you add the contract?

Then we merge!

**Comment by monperrus** [comment]
Thanks @andrewbwogi

**Pull request merged**